### ISSUE TITLE
Cannot assign ALTER PARTITION, because another ALTER PARTITION query was concurrently executed

### ISSUE THREAD

**Issue by Blackmorse** [ClickHouse/ClickHouse#30680]
### The problem

According to the documentation about the Synchronicity of ALTER Queries:

>  For replicated tables, the query just adds instructions for the appropriate actions to ZooKeeper, and the actions themselves are performed as soon as possible. However, the query can wait for these actions to be completed on all the replicas.

I expect that concurrent ALTER requests will make a queue and executed one after another with synchronization through Zookeeper, but instead of this I receive exceptions:

`DB::Exception: Cannot assign ALTER PARTITION, because another ALTER PARTITION query was concurrently executed.`

Clickhouse version 21.6.5.37

### Steps to reproduce
Clickhouse cluster _cluster_name_ with 2 Clickhouse servers (e.g. host1, host2) in one shard, Zookeeper.

```
CREATE TABLE default.destination ON CLUSTER cluster_name
(
    `a` UInt64,
    `b` String
)
ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/default/destination', '{replica}')
PARTITION BY b
ORDER BY a
SETTINGS index_granularity = 8192
```

host1:
```
CREATE TABLE default.source1 AS default.destination 
ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/default/source1', '{replica}')
PARTITION BY b
ORDER BY a
SETTINGS index_granularity = 8192;

INSERT INTO default.source1 SELECT *, '1' FROM numbers(100000000);
```

host2:
```
CREATE TABLE default.source2 AS default.destination 
ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/default/source2', '{replica}')
PARTITION BY b
ORDER BY a
SETTINGS index_granularity = 8192;

INSERT INTO default.source2 SELECT *, '1' FROM numbers(100000000);
```
(important: source1 and source2 are not the replicas. I use ReplicatedMergeTree for them to make MOVE PARTITION query works)

Then I create a script at one of the hosts:
```
#!/bin/bash

clickhouse-client --replication_alter_partitions_sync 0 --host host1 -q "alter table default.source1 MOVE PARTITION '1' to table default.destination" &

clickhouse-client --replication_alter_partitions_sync 0 --host host2 -q "alter table default.source2 MOVE PARTITION '1' to table default.destination" &
```
Error after the execution:
```
DB::Exception: Cannot assign ALTER PARTITION, because another ALTER PARTITION query was concurrently executed.
```

**Comment by amosbird** [comment]
It's expected after https://github.com/ClickHouse/ClickHouse/pull/24208 to avoid possible replica diverge. But we probably can do better (at least adding some internal retries)

**Comment by tavplubix** [comment]
Fixed in  #31010
